span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/ExcludeSpanRulesCache.java

```diff
@@ -80,9 +84,10 @@ public List<ExcludeSpanRule> load(@Nonnull ContextualKey<Void> key) {
         CACHE_NAME, excludeSpanRulesCache, Collections.emptyMap());
   }
 
-  public static synchronized ExcludeSpanRulesCache getInstance(Config config) {
+  static synchronized ExcludeSpanRulesCache getInstance(
```

Comment:
manually managed singletons are not a pattern we generally like, but pre-existing. Specifically the issue here though is that the arguments are ignored if the instance has already been created. This makes things like testing exceptionally difficult. It's not a blocker, but shouldn't be too hard - can we either consider removing this accessor, or if we really need it, making sure the returned instance matches the requested config/registry?

Reply:
this path is not being used by the tests - tests create their own mock cache instance and pass to the caller..

Reply:
It builds in a lot of assumptions - even if no test is triggering a path to INSTANCE today, all "regular" callers are assumed to call it with the same args. So everything may be set up such that it works correctly today, but there's no guarantee those assumptions will hold forever, and these bugs are annoying to track down. I know this one from experience, because we've made these exact same assumptions with the same pattern in other places and had to debug and fix it.

Anyway, as I mentioned it's pre-existing, so doesn't need to be fixed here - but could we leave a TODO to find an alternative approach that provides the context (a link to this thread is fine)?

Reply:
It's tricky to solve this without refactoring and using guice because we want to ensure single instance across all kafka processors.. adding a TODO for now..